From e23728e30ca99fb8dad7bec28aecae255862b7c7 Mon Sep 17 00:00:00 2001
From: Seungha Yang <seungha@centricular.com>
Date: Sat, 11 Apr 2020 03:11:35 +0900
Subject: [PATCH 36/60] kmssink: Add support for P010 and P016 formats

Support high bit-depth YUV formats

Part-of: <https://gitlab.freedesktop.org/gstreamer/gst-plugins-bad/-/merge_requests/1188>
(cherry picked from commit 09a8d2db05bff0526140187f68c39e68b706543f)
Signed-off-by: Jeffy Chen <jeffy.chen@rock-chips.com>
---
 sys/kms/gstkmsallocator.c |  6 ++++++
 sys/kms/gstkmsutils.c     | 20 ++++++++++++++++++++
 2 files changed, 26 insertions(+)

diff --git a/sys/kms/gstkmsallocator.c b/sys/kms/gstkmsallocator.c
index 916326b16..a0343922b 100644
--- a/sys/kms/gstkmsallocator.c
+++ b/sys/kms/gstkmsallocator.c
@@ -151,6 +151,12 @@ extrapolate_stride (const GstVideoFormatInfo * finfo, gint plane, gint stride)
     case GST_VIDEO_FORMAT_NV16:
     case GST_VIDEO_FORMAT_NV61:
     case GST_VIDEO_FORMAT_NV24:
+    case GST_VIDEO_FORMAT_P010_10LE:
+#ifdef GST_VIDEO_FORMAT_P010_10BE
+    case GST_VIDEO_FORMAT_P010_10BE:
+    case GST_VIDEO_FORMAT_P016_LE:
+    case GST_VIDEO_FORMAT_P016_BE:
+#endif
       estride = (plane == 0 ? 1 : 2) *
           GST_VIDEO_FORMAT_INFO_SCALE_WIDTH (finfo, plane, stride);
       break;
diff --git a/sys/kms/gstkmsutils.c b/sys/kms/gstkmsutils.c
index 90311f3fa..f8620c810 100644
--- a/sys/kms/gstkmsutils.c
+++ b/sys/kms/gstkmsutils.c
@@ -31,6 +31,8 @@
 
 #include "gstkmsutils.h"
 
+#define DRM_FORMAT_P010 DRM_FORMAT_NV12_10
+
 /* *INDENT-OFF* */
 static const struct
 {
@@ -49,6 +51,10 @@ static const struct
   DEF_FMT (XBGR8888, RGBx),
   DEF_FMT (BGR888, RGB),
   DEF_FMT (RGB888, BGR),
+  DEF_FMT (P010, P010_10LE),
+#ifdef DRM_FORMAT_P016
+  DEF_FMT (P016, P016_LE),
+#endif
 #else
   DEF_FMT (ARGB8888, ARGB),
   DEF_FMT (XRGB8888, xRGB),
@@ -56,6 +62,10 @@ static const struct
   DEF_FMT (XBGR8888, xBGR),
   DEF_FMT (RGB888, RGB),
   DEF_FMT (BGR888, BGR),
+  DEF_FMT (P010, P010_10BE),
+#ifdef DRM_FORMAT_P016
+  DEF_FMT (P016, P016_BE),
+#endif
 #endif
   DEF_FMT (UYVY, UYVY),
   DEF_FMT (YUYV, YUY2),
@@ -111,9 +121,15 @@ gst_drm_bpp_from_drm (guint32 drmfmt)
     case DRM_FORMAT_NV16:
       bpp = 8;
       break;
+    case DRM_FORMAT_P010:
+      bpp = 10;
+      break;
     case DRM_FORMAT_UYVY:
     case DRM_FORMAT_YUYV:
     case DRM_FORMAT_YVYU:
+#ifdef DRM_FORMAT_P016
+    case DRM_FORMAT_P016:
+#endif
       bpp = 16;
       break;
     case DRM_FORMAT_BGR888:
@@ -139,6 +155,10 @@ gst_drm_height_from_drm (guint32 drmfmt, guint32 height)
     case DRM_FORMAT_YUV422:
     case DRM_FORMAT_NV12:
     case DRM_FORMAT_NV21:
+    case DRM_FORMAT_P010:
+#ifdef DRM_FORMAT_P016
+    case DRM_FORMAT_P016:
+#endif
       ret = height * 3 / 2;
       break;
     case DRM_FORMAT_NV16:
-- 
2.20.1

